connect/runtime/src/main/java/org/apache/kafka/connect/util/LoggingContext.java

```diff
@@ -0,0 +1,142 @@
+         * implementations.
+         */
+        TASK("task"),
+
+        /**
+         * The {@link Parameters#CONNECTOR_SCOPE scope} value for committing offsets.
+         */
+        OFFSETS("offsets");
+
+        private final String text;
+        Scope(String value) {
+            this.text = value;
+        }
+
+        @Override
+        public String toString() {
+            return text;
+        }
+    }
+
+    /**
+     * Modify the current {@link MDC} logging context to set the {@link Parameters#CONNECTOR_NAME connector name} to the
+     * supplied name and the {@link Parameters#CONNECTOR_SCOPE scope} to {@link Scope#WORKER}.
+     *
+     * @param connectorName the connector name; may not be null
+     */
+    public static LoggingContext forConnector(String connectorName) {
+        LoggingContext context = new LoggingContext();
+        MDC.put(Parameters.CONNECTOR_NAME, connectorName);
+        MDC.remove(Parameters.CONNECTOR_TASK);
+        MDC.put(Parameters.CONNECTOR_SCOPE, Scope.WORKER.toString());
+        return context;
+    }
+
+    /**
+     * Modify the current {@link MDC} logging context to set the {@link Parameters#CONNECTOR_NAME connector name} and
+     * {@link Parameters#CONNECTOR_TASK task number} to the same values in the supplied {@link ConnectorTaskId}.
+     *
+     * @param id the connector task ID; may not be null
+     */
+    public static LoggingContext forTask(ConnectorTaskId id) {
+        LoggingContext context = new LoggingContext();
+        MDC.put(Parameters.CONNECTOR_NAME, id.connector());
+        MDC.put(Parameters.CONNECTOR_TASK, Integer.toString(id.task()));
+        MDC.put(Parameters.CONNECTOR_SCOPE, Scope.TASK.toString());
+        return context;
+    }
+
+    /**
+     * Modify the current {@link MDC} logging context to set the  {@link Parameters#CONNECTOR_NAME connector name},
+     * {@link Parameters#CONNECTOR_TASK task number} and to the same values in the supplied {@link ConnectorTaskId}, and to set
+     * the scope to {@link Scope#OFFSETS}.
+     *
+     * @param id the connector task ID; may not be null
+     */
+    public static LoggingContext forOffsets(ConnectorTaskId id) {
+        LoggingContext context = new LoggingContext();
+        MDC.put(Parameters.CONNECTOR_NAME, id.connector());
+        MDC.put(Parameters.CONNECTOR_TASK, Integer.toString(id.task()));
+        MDC.put(Parameters.CONNECTOR_SCOPE, Scope.OFFSETS.toString());
+        return context;
+    }
+
+    private final Map<String, String> previous;
+
+    private LoggingContext() {
+        previous = MDC.getCopyOfContextMap(); // may be null!
+    }
+
+    /**
+     * Close this logging context, restoring all {@link MDC} parameters back to the state just before this context was created.
+     */
+    @Override
+    public void close() {
+        MDC.clear();
+        if (previous != null && !previous.isEmpty()) {
+            MDC.setContextMap(previous);
```

Comment:
That was considered in an earlier version, and @mageshn suggested we do this.

Reply:
The only time a LoggingContext should be cleared is when a thread gets repurposed for a different reason. IIRC, the LoggingContext was not `Closeable` when I provided the original comment. If the general agreement is that it should be `Closeable` then we got find a way to manage the state then. IMO, I think it could become complicated trying to manage the state when the MDC is already doing in in threadLocal. One another interesting idea could be invoke close some how as part of the Plugin class loader swap, while not intuitive, it will ensure that we are clearing the context everytime its necessary. Thoughts?

Reply:
I think the source of the confusion arises from LoggingContext.close() calling MDC.clear. Removing specific keys instead would solve the problem.

Reply:
Actually, the purpose of `LoggingContext` is to essentially capture the MDC state such that it can be restored at a later time when `LoggingContext.close()` is called. The purpose is not just to manage the now one MDC parameter in which we're storing the Connect context. The `close()` method does call `MDC.clear()`, but then it also sets the MDC state to the same parameters that were found when the `LoggingContext` constructor was called.

Consider when this is used to set up a context under which a connector implementation method is called, and that connector implementation sets an MDC parameter. With the current implementation, when the logging context is closed, any changes to the MDC state will have been cleaned up properly, and the MDC state restored to exactly what it was when the context was first created.

Reply:
@wicknicks @rhauch @mageshn I think this behavior is acceptable. We may want to document this logging behavior somewhere so that nobody tries to set an `MDC` key in, e.g., `Connector.start()` and then refer back to it in `Connector.stop()`. However, at least from my POV it seems unreasonable and unintuitive to expect that to work in the first place. 

@rhauch I like the `shouldAllowNestedLoggingContexts` test you've written. As long as we agree that the behavior we're discussing here is acceptable, could you either add a few extra `MDC` keys inside one of the nested contexts and verify that they're cleared once the context is closed, or add a new test similar to `shouldAllowNestedLoggingContexts` that deals specifically with the wiping of extra `MDC` keys once a context is closed?

Reply:
I think I see @wicknicks's point, though, and I think the current implementation might pose a problem. Consider an implementation of `SourceTask` that sets up an MDC parameter in the `start(...)` method and remove it in the `stop(...)` method, but wants any log messages written by this task class to include that MDC parameter. See [this real-world example](https://github.com/debezium/debezium/blob/9b405aa84d8391a7ccf883184d6e8c67bac8bed3/debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java).

Connect currently runs each task in a separate thread via the `Worker.executorService`, and all of the task methods are called from within that thread. So I think it's safe to not restore the MDC state to exactly what it was when the LoggingContext was stopped, as suggested by @wicknicks.

However, when the task ends, it's possible the executor service will reuse the thread, so we need to clean that in `WorkerTask.run()`, just in case the thread's MDC parameters contain state from a prior task.